pkg/mcp/internal/queue.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"sync"
+	"time"
+
+	"istio.io/istio/pkg/mcp/env"
+)
+
+// TODO - this can eventually be moved under pkg/mcp/source once the new stack is
+// introduced. The code is temporarily added under pkg/mcp/internal so
+// we can share across package boundaries without exposing outside of the pkg/mcp
+// parent directory.
+
+// UniqueQueue is a specialized queue structure. It has the following properties:
+//
+// - Enqueuing an item already in the queue is idempotent. The first time an item
+//   is added to the queue determines its order in the queue until it is dequeued.
+//   Callers can safely update the queued items state while preserving it's place
+//   in the queue.
+//
+// - Enqueuing an item in the queue creates a schedule event. The caller can select
+//   over the Ready() channel to process this event and remove items from the queue.
+//
+// - The maximum queue depth is fixed.
+//
+// - The queue can be safely closed. The caller is responsible for checking the Done()
+//   state of the queue and should stop checking Ready() and invoking Dequeue() when
+//   the channel returned by Done() is closed.
+//
+// This is intended to be used by the MCP source/server packages for managing
+// per-type watch state.
+type UniqueQueue struct {
+	sync.Mutex
```

Comment:
Doesn't this expose the mutex to external callers?

Reply:
Yes. This was less of a problem when the implementation lived in the same package as the caller (pkg/mcp/server/). Making it private makes more sense now that it's in its own package.